### PR TITLE
Improve memory consumption of condensation losses

### PR DESCRIPTION
**Pull request by jnpark3**
Pull request in resolving issue #293.

Background loss calculation was updated to require less memory usage while computing the indices of alpha particles. Potential loss calculation was updated by adding the '_fast_condensation_loss' function, which uses the updated alpha index search methods and radius graphs to significantly decrease the memory usage of computing condensation loss. This function uses an approximate method to calculate repulsive loss, as torch_cluster's 'radius_graph' requires an upper limit 'max_neighbors' to the number of edges constructed originating from each point. This limitation should not significantly change the characteristic of the loss function for sufficiently large values of 'max_neighbors'. When 'max_neighbors' is set to <=0 while setting up 'PotentialLoss' object, the original condensation loss function is used.

Comments added to explain in detail the workings of this function.

### CODE REVIEW

File: src/gnn_tracking/metrics/losses.py [hunk 1]
```diff
@@ -226,21 +227,121 @@ def _condensation_loss(
...
+    alpha_hits_filter = torch.zeros(
+        len(particle_id), dtype=bool, device=x.device
+    ).scatter_(0, alphas, 1)
+    # indices of all non-CPs
+    non_alpha_indices = torch.arange(len(particle_id), device=x.device)[
+        ~alpha_hits_filter
+    ]
+
+    # for each non-CP hit, the index of the corresponding CP
+    alpha_indices = _alphas[
```
Comment (klieret) [inline src/gnn_tracking/metrics/losses.py:287]: It seems very strange that this includes a "noise CP", i.e., I think this should be `alphas`, not `_alphas`

File: src/gnn_tracking/metrics/losses.py [hunk 2]
```diff
@@ -226,21 +227,121 @@ def _condensation_loss(
...
+    _radius_edges = radius_graph(
+        x=x, r=radius_threshold, max_num_neighbors=max_num_neighbors, loop=False
+    )
+    # Now filter out everything that doesn't include a CP or connects two hits of the
+    # same particle
+    _to_cp = torch.isin(_radius_edges[0], alphas)
+    _is_repulsive = particle_id[_to_cp[0]] != particle_id[_to_cp[1]]
+    repulsion_edges = _to_cp[:, _is_repulsive & _to_cp]
+
+    # -- 3. Edges for attractive loss --
```
Comment (klieret) [inline src/gnn_tracking/metrics/losses.py:276]: couldn't we simplify this code by allowing loops for the CPs? The distances to itself would be 0 anyway, so it would make no difference, right?

File: src/gnn_tracking/metrics/losses.py [hunk 3]
```diff
@@ -226,21 +227,121 @@ def _condensation_loss(
...
+    unmasked_attraction_edges = (
+        torch.arange(len(particle_id), device=x.device).unsqueeze(0).repeat(2, 1)
+    )
+    unmasked_attraction_edges[1, ~alpha_hits_filter] = alpha_indices
+
+    # Apply mask to attraction edges
+    noisy_attraction_edges = unmasked_attraction_edges[:, mask]
+
+    # Filter out noise edges from attraction edges
+    attraction_edges = noisy_attraction_edges[
```
Comment (klieret) [inline src/gnn_tracking/metrics/losses.py:301]: the mask will already exclude noise (`pt > 0` excludes noise if you look at the mask definition), so this isn't needed

**Review by klieret — CHANGES_REQUESTED** [review]

**Comment by jnpark3** [comment]
pre-commit.ci autofix

File: src/gnn_tracking/metrics/losses.py (hunk 1 above)
Comment (jnpark3) [reply src/gnn_tracking/metrics/losses.py:287]: The "noise CP" is important for torch.searchsorted operation. If the noise cp were not there, the indices generated by this operation get shifted down by one, and it makes it impossible to differentiate between noise hits and hits with the lowest positive particle id (they both get labeled as 0).

File: src/gnn_tracking/metrics/losses.py [hunk 4]
```diff
@@ -264,22 +360,36 @@ def forward(
...
+        if self.max_neighbors <= 0:
+            return _condensation_loss(
+                beta=beta,
+                x=x,
+                particle_id=particle_id,
+                mask=mask,
+                q_min=self.q_min,
+                radius_threshold=self.radius_threshold,
+            )
+        else:
```
Comment (klieret) [inline src/gnn_tracking/metrics/losses.py:372]: (ruff complains that you can simply remove the `else` here)

**Review by klieret — APPROVED** [review]
